src/main/java/org/elasticsearch/search/aggregations/reducers/BucketHelpers.java

```diff
@@ -0,0 +1,187 @@
+     * @param bucket     A specific bucket that a value needs to be extracted from.  This bucket should be present
+     *                   in the <code>histo</code> parameter
+     * @param aggPath    The path to a particular value that needs to be extracted.  This path should point to a metric
+     *                   inside the <code>bucket</code>
+     * @param gapPolicy  The gap policy to apply if empty buckets are found
+     * @return           The value extracted from <code>bucket</code> found at <code>aggPath</code>
+     */
+    public static Double resolveBucketValue(InternalMultiBucketAggregation<?, ? extends InternalMultiBucketAggregation.Bucket> agg,
+            InternalMultiBucketAggregation.Bucket bucket, String aggPath, GapPolicy gapPolicy) {
+        List<String> aggPathsList = AggregationPath.parse(aggPath).getPathElementsAsStringList();
+        return resolveBucketValue(agg, bucket, aggPathsList, gapPolicy);
+    }
+
+    public static Double resolveBucketValue(InternalMultiBucketAggregation<?, ? extends InternalMultiBucketAggregation.Bucket> agg,
+            InternalMultiBucketAggregation.Bucket bucket, List<String> aggPathAsList, GapPolicy gapPolicy) {
+        try {
+            Object propertyValue = bucket.getProperty(agg.getName(), aggPathAsList);
+            if (propertyValue == null) {
+                throw new AggregationExecutionException(DerivativeParser.BUCKETS_PATH.getPreferredName()
+                        + " must reference either a number value or a single value numeric metric aggregation");
+            } else {
+                double value;
+                if (propertyValue instanceof Number) {
+                    value = ((Number) propertyValue).doubleValue();
+                } else if (propertyValue instanceof InternalNumericMetricsAggregation.SingleValue) {
+                    value = ((InternalNumericMetricsAggregation.SingleValue) propertyValue).value();
+                } else {
+                    throw new AggregationExecutionException(DerivativeParser.BUCKETS_PATH.getPreferredName()
+                            + " must reference either a number value or a single value numeric metric aggregation");
+                }
+                // doc count never has missing values so gap policy doesn't apply here
+                boolean isDocCountProperty = aggPathAsList.size() == 1 && "_count".equals(aggPathAsList.get(0));
+                if (Double.isInfinite(value) || Double.isNaN(value) || (bucket.getDocCount() == 0 && !isDocCountProperty)) {
+                    switch (gapPolicy) {
+                    case INSERT_ZEROS:
+                        return 0.0;
+                    case IGNORE:
+                    default:
+                        return Double.NaN;
+                    }
+                } else {
+                    return value;
+                }
+            }
+        } catch (InvalidAggregationPathException e) {
+            return null;
```

Comment:
Should we add a debug log here then? it still needs to return null as a bucket may not have an entry for a particular aggregation name